### PR TITLE
When NAND is damaged, show title names from save files

### PR DESCRIPTION
**Pull request by JosJuice**
The earlier code always tried to use TitleDatabase for getting title names, but that didn't work for disc-based games, because there was no way to get the maker ID.

### CODE REVIEW

File: Source/Core/Common/StringUtil.cpp [hunk 1]
```diff
@@ -576,3 +577,11 @@ std::string UTF16ToUTF8(const std::wstring& input)
 }
 
 #endif
+
+std::string UTF16BEToUTF8(const char16_t* str, size_t max_size)
+{
+  const char16_t* str_end = std::find(str, str + max_size, '\0');
+  std::wstring result(static_cast<size_t>(str_end - str), '\0');
+  std::transform(str, str_end, result.begin(), (u16(&)(u16))Common::swap16);
```
Comment (lioncash) [inline Source/Core/Common/StringUtil.cpp]: `static_cast`

File: Source/Core/DiscIO/Volume.cpp [hunk 2]
```diff
@@ -73,11 +35,8 @@ std::map<Language, std::string> Volume::ReadWiiNames(const std::vector<u8>& data
     size_t name_end = name_start + NAME_BYTES_LENGTH;
     if (data.size() >= name_end)
     {
-      u16* temp = (u16*)(data.data() + name_start);
-      std::wstring out_temp(NAME_STRING_LENGTH, '\0');
-      std::transform(temp, temp + out_temp.size(), out_temp.begin(), (u16(&)(u16))Common::swap16);
-      out_temp.erase(std::find(out_temp.begin(), out_temp.end(), 0x00), out_temp.end());
-      std::string name = UTF16ToUTF8(out_temp);
+      std::string name = UTF16BEToUTF8(reinterpret_cast<const char16_t*>(data.data() + name_start),
```
Comment (lioncash) [inline Source/Core/DiscIO/Volume.cpp]: This dereferencing the pointer result of this cast is undefined behavior by the way (despite being pre-existing).

File: Source/Core/DiscIO/WiiSaveBanner.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#include <vector>
+
+#include "Common/CommonTypes.h"
+
+namespace DiscIO
+{
+class WiiSaveBanner
+{
+public:
+  WiiSaveBanner(u64 title_id);
```
Comment (lioncash) [inline Source/Core/DiscIO/WiiSaveBanner.h]:
```cpp
explicit WiiSaveBanner(u64 title_id);
explicit WiiSaveBanner(const std::string& path);
```

**Comment by JosJuice** [comment]
Rebased.

**Pull request merged**